Full CRUD of departments, sectors, and users

Snail51 opened the pull request:
**BASED ON #15 (NOT main), ADDRESS #15 BEFORE THIS** 

Addresses #3 , #4 
Slight Improvements to #5 
Addresses #13 for these objects (users, sectors, departments)

Full implementation of CRUD for the Users, Sectors, and Departments objects
All routes and functionality have been confirmed to be working on these direct pages.

Note: "Staff Count" field on Departments and Sectors remains unimplemented and is always a "-"

Note: User editing is slightly broken, as removing a user's sector does not clear their department, allowing them to have a department but not a sector. This wont really happen in production but should probably be addressed at some point.

Snail51 committed:
standardize naming convention `departments`

all calls to views now use `departments` (with an s) instead of `department`

Snail51 committed:
Merge branch 'main' into CRUD_departments

Snail51 committed:
restore stashed changes

Snail51 committed:
impliment department creation & update

slightly crude implimentation of departments and supporting sector infrastructure

Snail51 committed:
crude impliment delete (hard)

deltion fo departments now possible, currently using hard deletes

Snail51 committed:
experiemnts with routes

Snail51 committed:
fix user routes

i now actually understand routes and was able to fix them. the previous push had a collision that caused users.update to map to users.delete

all forms bust use method=post in the <> but then must override to @method('patch') (for example) to become the new type

i also now understand how Route::resources works and have applied it properly

Snail51 committed:
Merge branch 'admin-delete-soft' into CRUD_departments

Snail51 committed:
fix users.delete_confirm route

cannot use "." in route name because that implies directory traversal, must use something else instead (in this case)

also  had to rename the delete view to match that new route name

Snail51 committed:
Merge branch 'admin-delete-soft' into CRUD_departments

Snail51 committed:
fix department routes to match standard Route::resource

Snail51 committed:
department SCRUM confirmed working

all operations behave as expected

Snail51 committed:
fully impliment + test sectors CRUD

full correct CRUD for sectors, confirmed working.

.show and .edit pages a little bare-bones but are fully sufficient.

i have also tested what happens if a sector/department is deleted. cascade occurs but users are not effected (beyond those attributes being set to null)

note: "Staff Count" fields for both departments and sectors are not yet implemented and always display "-"

note: user editing is somewhat broken, as removing a user's sector does not clear their department, allowing them to have a department but not a sector

Snail51 committed:
improve error handling

wrap all my CRUD pages (users, departments, sectors) in @auth @endauth blocks to help with error handling. this is okay because we dont want non-signed in users to see these pages anyway.

also fixed user controller to use findOrFail instead of find

Snail51 committed:
401 aborts

instead of redirecting to `return view('dashboard');` when the viewer is not an admin, instead use `abort(401);`. this shouldnt actually matter, as the auth middleware should handle this, but it's more correct for edge cases.

ALSO: there was a typo in the @endauth directive in `../views/users/index.blade.php` which was fixed

drraccoony commented on app/Http/Controllers/DepartmentController.php:
```diff
@@ -15,26 +25,60 @@ class DepartmentController extends Controller
+            'sector_id' => ['required','integer','exists:sectors,id']     // Ensure 'sector' is a valid integer and exists in the sectors table
+        ]);
+
+        // Programmatically determine the department ID of this new department
+        $lastDepartmentId = Department::latest()->first()->id ?? -1; //if no other department exists, start at -1+1 (0)
+        $newDepartmentId = $lastDepartmentId + 1;
+        $validated['id'] = $newDepartmentId; 
```
This isn't needed. primary key ID's with laravel are automatically incrementally generated for us. We dont need to do this. But I see what you were doing.

Snail51 replied:
resolved in 59ff09f7743b739de37cce6a45b1e3a7053d17fb

drraccoony commented on app/Http/Controllers/DepartmentController.php:
```diff
@@ -51,28 +96,86 @@ public function show(string $id)
-        return view('departments.edit', [
-            'department' => $department
-        ]);
+        if($request->user()->isAdmin())
```
You shouldn't need to do request for the user, remember you should be using `Auth::`. A request can be manipulated to fool the controller into thinking its an admin, whereas the Auth facade is handled on the backend for us with session management.

Refactor this to use `Auth::user`.

Snail51 replied:
resolved in 59ff09f7743b739de37cce6a45b1e3a7053d17fb

drraccoony commented on app/Http/Controllers/DepartmentController.php:
```diff
@@ -51,28 +96,86 @@ public function show(string $id)
+    public function delete(Request $request, string $id): View
     {
-        //
+        if($request->user()->isAdmin())
```
Refactor to use `Auth::user` facade.

Snail51 replied:
resolved in 59ff09f7743b739de37cce6a45b1e3a7053d17fb

drraccoony commented on app/Http/Controllers/SectorController.php:
```diff
@@ -24,30 +24,39 @@ public function index()
+    public function create(Request $request)
     {
-        return view('sectors.create');
+        if($request->user()->isAdmin())
```
Refactor to use `Auth::user`.

Snail51 replied:
resolved in 59ff09f7743b739de37cce6a45b1e3a7053d17fb

drraccoony commented on app/Http/Controllers/SectorController.php:
```diff
@@ -57,35 +66,88 @@ public function store(Request $request)
-        return view('sectors.edit', [
-            'sector' => $sector
-        ]);
+        if($request->user()->isAdmin())
```
Refactor to use `Auth::user`.

Snail51 replied:
resolved in 59ff09f7743b739de37cce6a45b1e3a7053d17fb

drraccoony commented on app/Http/Controllers/SectorController.php:
```diff
@@ -57,35 +66,88 @@ public function store(Request $request)
+    public function destroy(Request $request, string $id)
     {
-        //
+        if($request->user()->isAdmin())
```
Refactor to use `Auth::user`

Snail51 replied:
resolved in 59ff09f7743b739de37cce6a45b1e3a7053d17fb

drraccoony commented on app/Http/Controllers/UserController.php:
```diff
@@ -172,8 +172,8 @@ public function delete(Request $request, string $id): View
     {
         if($request->user()->isAdmin())
```
This should be `Auth::user`.

Snail51 replied:
resolved in 59ff09f7743b739de37cce6a45b1e3a7053d17fb

drraccoony commented on routes/web.php:
If `Route::middleware('auth')->group(function () {` is being called multiple times, it wouldn't hurt to nest all routes that depend on this middleware once. Just to clean it up a little.

Snail51 replied:
resolved in 9c6818f8ae52ff831beb2cc25623d003dcb30c5a

drraccoony reviewed (changes requested):
Good work on the crud calls, just some minor adjustments and its good.

Snail51 committed:
use Auth:: directive for controller verification

- all instances where `$request->user()->isAdmin()` were used are now `Auth::check() && Auth::user()->isAdmin()`

- lingering redirects were replaced with `abort(401)`

- removed unnecrssary lines in `DepartmentController.php` doing explicit incrimentation of department ids

Snail51 committed:
improve clarity of `../routes/web.php`

remove duplicate auth middleware declarations
